kuka_rsi_simulator/kuka_rsi_simulator/simulator.py

```diff
@@ -37,38 +38,48 @@ def _create_rsi_xml_rob(self):
         root = ET.Element('Rob', {'TYPE': 'KUKA'})
         ET.SubElement(root, 'AIPos', {'A1': str(self._q[0]), 'A2': str(self._q[1]), 'A3': str(self._q[2]),
                                       'A4': str(self._q[3]), 'A5': str(self._q[4]), 'A6': str(self._q[5])})
-        ET.SubElement(root, 'AIVel', {'A1': str(self._dq[0]), 'A2': str(self._dq[1]), 'A3': str(self._dq[2]),
-                                      'A4': str(self._dq[3]), 'A5': str(self._dq[4]), 'A6': str(self._dq[5])})
+        ET.SubElement(root, 'AIVel', {'A1': str(self._qd[0]), 'A2': str(self._qd[1]), 'A3': str(self._qd[2]),
+                                      'A4': str(self._qd[3]), 'A5': str(self._qd[4]), 'A6': str(self._qd[5])})
         ET.SubElement(root, 'FT', {'Fx': str(self._wrench[0]), 'Fy': str(self._wrench[1]), 'Fz': str(self._wrench[2]),
                                    'Tx': str(self._wrench[3]), 'Ty': str(self._wrench[4]), 'Tz': str(self._wrench[5])})
+        ET.SubElement(root, 'DigOut', {"digital_output_0": str(self._digital_output_0)}),
         ET.SubElement(root, 'Delay', {'D': str(self._timeout_count)})
         ET.SubElement(root, 'SENSOR').text = "4.0"
         ET.SubElement(root, 'IPOC').text = str(self._ipoc)
         return ET.tostring(root)
 
     def _parse_rsi_xml_sen(self, data):
         root = ET.fromstring(data)
+        if self._ipoc != int(root.find('IPOC').text):
+            raise RuntimeError(f"IPOC incorrect, expected {self._ipoc} got {root.find('IPOC').text}")
+        self._ipoc += 1
+        if root.find('Stop').text == "1":
+            self.get_logger().info("RSI stopped")
+            exit(0)
         axis = root.find('Axis').attrib
         desired_joint_correction = np.array([axis['A1'], axis['A2'], axis['A3'],
                                              axis['A4'], axis['A5'], axis['A6']]).astype(np.float64)
-        if self._ipoc != int(root.find('IPOC').text):
-            raise
+        self._digital_output_0 = root.find('DigOut').attrib['digital_output_0']
         return desired_joint_correction
 
     def timer_callback(self):
+        if self._timeout_count >= 100:
+            raise RuntimeError("Timeout count reached")
         try:
-            msg = self._create_rsi_xml_rob()
-            self._s.sendto(msg, (self._host, self._port))
+            if not self._timeout_count:
+                msg = self._create_rsi_xml_rob()
+                self._s.sendto(msg, (self._host, self._port))
             recv_msg, _ = self._s.recvfrom(1024)
-            self._dq = self._parse_rsi_xml_sen(recv_msg)
-            self._q += self._dq
+            dq = self._parse_rsi_xml_sen(recv_msg)
+            q_new = self._q + dq
```

Comment:
```suggestion
            qd = self._parse_rsi_xml_sen(recv_msg)
            q_new = self._q + qd
```
not important nit

Reply:
This was a concious choice because the commands are actually deltas in positions and not velocities